### PR TITLE
Add (unstable) FnBox trait as a nicer replacement for `Thunk`. 

### PR DESCRIPTION
**Pull request by nikomatsakis**
Add (unstable) FnBox trait as a nicer replacement for `Thunk`. The doc comment includes a test that also shows how it can be used.

f? @aturon 
f? @alexcrichton

### CODE REVIEW

**Comment by nikomatsakis** [comment]
Hmm, encountered an obstacle. Investigating now.

**Comment by nikomatsakis** [comment]
OK, so, there is an obstacle that I'm not sure how best to rectify, though it's also not a strict blocker. Just make the `FnBox` story less good than I hoped it was. (Of course, `FnBox` has always been seen as a kind of temporary work-around.)

If you have a type like `x: Box<FnBox(&mut Foo)>` (note the bound region there), you can't call it with `x()` notation, you must write `x.call_box((arg,))`. The reason is that the impl is for `Box<FnBox<A>>` which is less polymorphic than Box<for<'a> FnBox<(&'a mut Foo,)>>`. The compiler's normal "instantiate polymorphic region" logic doesn't help here, because the call is being redirected through an impl, which requires that `Box<FnBox<A>> == Box<for<'a> FnBox<(&'a mut Foo,)>>` for some `A` that is just not possible.

There are two workarounds. You can manually upcast:

```rust
let x: Box<FnBox<(&mut Foo,), Output=()>> = x;
```

which works because `Box` is covariant, or just use `call_box` directly. 

(Interestingly, this *would* be a kind of use case for variance on traits (that I just opened a PR to remove) except that `FnBox` winds up being invariant due to associated types, kind of proving my point that it's not buying us as much as it should.)

File: src/libtest/lib.rs [hunk 1]
```diff
@@ -958,7 +959,7 @@ pub fn run_test(opts: &TestOpts,
         }
         DynMetricFn(f) => {
             let mut mm = MetricMap::new();
-            f.invoke(&mut mm);
+            f.call_box((&mut mm,)); // TODO unfortunate
```
Comment (alexcrichton) [inline src/libtest/lib.rs]: I think this may need to be "FIXME" to get past `make tidy`

File: src/libstd/thunk.rs [hunk 2]
```diff
@@ -12,45 +12,9 @@
...
 
-use alloc::boxed::Box;
+use alloc::boxed::{Box, FnBox};
 use core::marker::Send;
-use core::ops::FnOnce;
 
-pub struct Thunk<'a, A=(),R=()> {
-    invoke: Box<Invoke<A,R>+Send + 'a>,
-}
+pub type Thunk<'a, A=(), R=()> =
```
Comment (alexcrichton) [inline src/libstd/thunk.rs]: Could you add a `#[deprecated]` tag to the module as well?

**Comment by alexcrichton** [comment]
Nice! r=me with a few nits

I'm ok with having to use unstable syntax to use an unstable feature :)

**Comment by aturon** [comment]
LGTM as well. We can discuss the stabilization story at some point during the beta period.

**Pull request merged**